ops/charm.py

```diff
@@ -263,60 +339,150 @@ class RelationCreatedEvent(RelationEvent):
+    RelationBrokenEvent will fire to signal that the relationship has
+    been fully terminated.
+
+    The RelationBrokenEvent hook indicates that the current relation
+    is no longer valid, and that the charm’s software must be
+    configured as though the relation had never existed. It will only
+    be called after every necessary RelationDepartedEvent hook has
+    been run. If the RelationBrokenEvent hook is being executed, it is
+    gauranteed that no remote units are currently known locally.
     """
 
 
 class StorageEvent(HookEvent):
-    """Base class representing Storage related events."""
+    """Base class representing Storage related events.
+
+    Juju can provide a variety of storage to charms. The charms can
+    define several different types of storage that are allocated from
+    Juju. Changes in state of storage trigger sub-types of
+    StorageEvent hooks.
+    """
 
 
 class StorageAttachedEvent(StorageEvent):
     """Represents the `storage-attached` hook from Juju.
 
     Called when new storage is available for the charm to use.
+
+    StorageAttachedEvent hook allows the charm to run code when
+    storage has been added. The StorageAttachedEvent hooks will be run
+    before the install hook, so that the installation routine may use
+    the storage. The name prefix of this hook will depend on the
+    storage key defined in the `metadata.yaml` file.
     """
 
 
 class StorageDetachingEvent(StorageEvent):
     """Represents the `storage-detaching` hook from Juju.
 
-    Called when storage a charm has been using is going away.
+    Called when storage a charm has been using, is going away.
+
+    StorageDetachingEvent hook allows the charm to run code before
+    storage is removed. The StorageDetachingEvent hooks will be run
+    before storage is detached, and always before the stop hook is
+    run, to allow the charm to gracefully release resources before
+    they are removed and before the unit terminates. The name prefix
+    of the hook will depend on the storage key defined in the
+    `metadata.yaml` file.
     """
 
 
 class CharmEvents(ObjectEvents):
-    """The events that are generated by Juju in response to the lifecycle of an application."""
+    """Events generated by Juju pertaining to application lifecycle.
+
+    This class is used to create an event descriptor (`self.on`)
+    attribute for a Charm class that inherits from CharmBase. The
+    event descriptor may be used to setup event handlers for
+    corresponding events.
+
+    By default the following events will be provided through
+    CharmBase:
+
+    ```
+    self.on.install
+    self.on.start
+    self.on.remove
+    self.on.update_status
+    self.on.config_changed
+    self.on.upgrade_charm
+    self.on.pre_series_upgrade
+    self.on.post_series_upgrade
+    self.on.leader_elected
+    self.on.collect_metrics
+    ```
+
+    In addition to these named relation and storage events may also be
```

Comment:
This first sentence seems to be missing something. How is this as an alternative:

"In addition to named relation and storage events, it is possible to define named events in `metadata.yaml`."

Reply:
The intention was to explain that relations and storage resources defined in `metadata.yaml` lead to eponymous events. If this was not clear then perhaps it needs to be reworded. It will help if you can clarify what you found to be unclear.

Reply:
Looking back at this, it reads properly with a comma. "In addition to these, named relation and storage events ..."

I was reading it without pause and it didn't make sense. Now it does.